forge-game/src/main/java/forge/game/GameAction.java

```diff
@@ -205,6 +205,7 @@ private Card changeZone(final Zone zoneFrom, Zone zoneTo, final Card c, Integer
             copied.setGameTimestamp(c.getGameTimestamp());
 
             if (zoneTo.is(ZoneType.Stack)) {
+                copied.setCastFrom(zoneFrom);
```

Comment:
if you need it earlier this should become obsolete:
https://github.com/Card-Forge/forge/blob/48a5046d9371b5eda6e9f709e2fe645e634dff55/forge-game/src/main/java/forge/game/GameAction.java#L829-840

Reply:
yeah that might be cleaned up, i think
needs to be tested